File: internal/tui/toolview.go
package tui

import (
	"encoding/json"
	"fmt"
	"strings"
)

// primaryArgKeys is the field a built-in tool's one-line summary should show.
// Names come from the tool schemas, not from anyone's example query.
var primaryArgKeys = map[string][]string{
	"exec":          {"command"},
	"web_search":    {"query"},
	"web_fetch":     {"url"},
	"read":          {"file_path"},
	"write":         {"file_path"},
	"edit":          {"file_path"},
	"ls":            {"path"},
	"tree":          {"path"},
	"glob":          {"pattern", "path"},
	"grep":          {"pattern", "path"},
	"python_runner": {"code"},
	"screenshot":    {"path"},
	"spawn_agent":   {"role"},
	"send_message":  {"text"},
	"resume_agent":  {"agent_id"},
	"wait_agents":   {},
	"close_agent":   {"agent_id"},
}

// summariseToolArgs turns a tool's raw argument string into the one line a
// human should see. JSON envelopes become the command / query / path; plain
// text is left alone so older `read(notes.md)` notifications still render.
func summariseToolArgs(name, args string) string {
	args = strings.TrimSpace(args)
	if args == "" {
		return ""
	}
	obj, ok := parseObject(args)
	if !ok {
		return flatten(args)
	}
	if keys, known := primaryArgKeys[name]; known {
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			if s := asText(obj[k]); s != "" {
				parts = append(parts, s)
			}
		}
		if len(parts) > 0 {
			return strings.Join(parts, " · ")
		}
		return ""
	}
	for _, k := range []string{"command", "query", "url", "file_path", "path", "pattern", "code", "text"} {
		if s := asText(obj[k]); s != "" {
			return s
		}
	}
	for _, v := range obj {
		if s := asText(v); s != "" {
			return s
		}
	}
	return flatten(args)
}

type toolView struct {
	failed bool
	error  string
	body   string
}

// viewToolResult parses a built-in tool's payload the same way the desktop UI
// does: exec stdout (or a failed flag), search hits, or the error: prefix.
func viewToolResult(name, raw string) toolView {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return toolView{}
	}
	if prefix, ok := errorPrefix(raw); ok {
		return toolView{failed: true, error: prefix}
	}
	if name == "exec" || name == "python_runner" {
		if run, ok := parseRunResult(raw); ok {
			failed := run.failed || (run.exitCode != nil && *run.exitCode != 0)
			body := joinNonEmpty(run.stdout, run.stderr)
			errText := run.err
			if failed && errText == "" {
				code := "?"
				if run.exitCode != nil {
					code = fmt.Sprintf("%d", *run.exitCode)
				}
				errText = "exit " + code
			}
			return toolView{failed: failed, error: errText, body: body}
		}
	}
	if name == "web_search" {
		if hits := parseSearchHits(raw); len(hits) > 0 {
			return toolView{body: strings.Join(hits, "\n")}
		}
	}
	return toolView{body: raw}
}

func (v toolView) display() string {
	if v.error != "" && v.body != "" {
		return v.error + "\n" + v.body
	}
	if v.error != "" {
		return v.error
	}
	return v.body
}

type runResult struct {
	stdout   string
	stderr   string
	failed   bool
	exitCode *int
	err      string
}

func parseRunResult(raw string) (runResult, bool) {
	obj, ok := parseObject(raw)
	if !ok {
		return runResult{}, false
	}
	if _, hasOut := obj["stdout"]; !hasOut {
		if _, hasErr := obj["stderr"]; !hasErr {
			if _, hasCode := obj["exit_code"]; !hasCode {
				if _, hasFail := obj["failed"]; !hasFail {
					return runResult{}, false
				}
			}
		}
	}
	run := runResult{
		stdout: asText(obj["stdout"]),
		stderr: asText(obj["stderr"]),
		failed: obj["failed"] == true,
		err:    asText(obj["error"]),
	}
	if n, ok := asInt(obj["exit_code"]); ok {
		run.exitCode = &n
	}
	return run, true
}

func parseSearchHits(raw string) []string {
	var val any
	if err := json.Unmarshal([]byte(raw), &val); err != nil {
		return nil
	}
	rows := asArray(val)
	if rows == nil {
		if obj, ok := val.(map[string]any); ok {
			rows = asArray(obj["results"])
			if rows == nil {
				rows = asArray(obj["items"])
			}
			if rows == nil {
				rows = asArray(obj["data"])
			}
		}
	}
	if rows == nil {
		return nil
	}
	hits := make([]string, 0, len(rows))
	for _, row := range rows {
		obj, ok := row.(map[string]any)
		if !ok {
			continue
		}
		title := firstText(obj, "title", "name")
		url := firstText(obj, "url", "link", "href")
		snippet := firstText(obj, "snippet", "summary", "body", "content", "description")
		if title == "" && url == "" && snippet == "" {
			continue
		}
		line := title
		if line == "" {
			line = url
		}
		if url != "" && url != title {
			line += " " + url
		}
		if snippet != "" {
			line += " — " + snippet
		}
		hits = append(hits, line)
	}
	return hits
}

func errorPrefix(raw string) (string, bool) {
	t := strings.TrimSpace(raw)
	if len(t) >= 6 && strings.EqualFold(t[:6], "error:") {
		return strings.TrimSpace(t[6:]), true
	}
	return "", false
}

func parseObject(raw string) (map[string]any, bool) {
	var val any
	if err := json.Unmarshal([]byte(raw), &val); err != nil {
		return nil, false
	}
	obj, ok := val.(map[string]any)
	return obj, ok
}

func asArray(v any) []any {
	a, ok := v.([]any)
	if !ok {
		return nil
	}
	return a
}

func asText(v any) string {
	switch t := v.(type) {
	case string:
		return strings.TrimSpace(t)
	case float64:
		if t == float64(int64(t)) {
			return fmt.Sprintf("%d", int64(t))
		}
		return fmt.Sprintf("%v", t)
	case bool:
		if t {
			return "true"
		}
		return "false"
	default:
		return ""
	}
}

func asInt(v any) (int, bool) {
	switch t := v.(type) {
	case float64:
		return int(t), true
	case json.Number:
		n, err := t.Int64()
		return int(n), err == nil
	case string:
		n := 0
		if _, err := fmt.Sscanf(strings.TrimSpace(t), "%d", &n); err == nil {
			return n, true
		}
	}
	return 0, false
}

func firstText(obj map[string]any, keys ...string) string {
	for _, k := range keys {
		if s := asText(obj[k]); s != "" {
			return s
		}
	}
	return ""
}

func joinNonEmpty(parts ...string) string {
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if strings.TrimSpace(p) != "" {
			out = append(out, p)
		}
	}
	return strings.Join(out, "\n")
}

func flatten(s string) string {
	return strings.Join(strings.Fields(s), " ")
}

File: frontend/src/lib/tool-view.ts
/** How a built-in tool call should read on one line, and how its payload
 *  should look once expanded. Field names come from the tool schemas — not
 *  from anyone's example query. */

const PRIMARY: Record<string, string[]> = {
  exec: ["command"],
  web_search: ["query"],
  web_fetch: ["url"],
  read: ["file_path"],
  write: ["file_path"],
  edit: ["file_path"],
  ls: ["path"],
  tree: ["path"],
  glob: ["pattern", "path"],
  grep: ["pattern", "path"],
  python_runner: ["code"],
  screenshot: ["path"],
  spawn_agent: ["role"],
  send_message: ["text"],
  resume_agent: ["agent_id"],
  wait_agents: [],
  close_agent: ["agent_id"],
  memory: ["action", "content", "old_text"],
  skill_view: ["name"],
  skill_manage: ["action", "name"],
}

export type SearchHit = { title: string; url: string; snippet: string }

export type ToolView = {
  summary: string
  failed: boolean
  error?: string
  body: string
  hits?: SearchHit[]
}

export function summariseToolCall(name: string, args: string): string {
  const obj = parseObject(args)
  if (name === "memory" || name === "skill_manage" || name === "skill_view") {
    return flatten(summariseMemoryCall(name, obj, args))
  }
  if (!obj) return flatten(args)
  const keys = PRIMARY[name]
  if (keys) {
    const parts = keys.map((k) => asText(obj[k])).filter(Boolean)
    if (parts.length) return flatten(parts.join(" · "))
    return ""
  }
  for (const k of ["command", "query", "url", "file_path", "path", "pattern", "code", "text"]) {
    const v = asText(obj[k])
    if (v) return flatten(v)
  }
  for (const v of Object.values(obj)) {
    const t = asText(v)
    if (t) return flatten(t)
  }
  return flatten(args)
}

/** The raw exec invocation, newlines kept. The one-line summary flattens
 *  this; the expanded body must not, or a heredoc becomes unreadable. */
export function execCommand(args: string): string {
  const obj = parseObject(args)
  if (obj) return asText(obj.command)
  return args.trim()
}

/** The collapsed tool row is what a human sees without opening it. A refused
 *  memory write that only shows the note being stored looks like a mystery;
 *  the refusal is the thing they need. */
export function toolRowSummary(view: ToolView): string {
  if (!view.failed || !view.error) return view.summary
  const err = flatten(view.error)
  if (!view.summary) return err
  if (view.summary.includes(err)) return view.summary
  return flatten(`${view.summary} · ${err}`)
}

export function viewTool(name: string, args: string, result?: string, flagged?: boolean): ToolView {
  const summary = summariseToolCall(name, args)
  if (result === undefined) {
    return { summary, failed: Boolean(flagged), body: "" }
  }
  const prefix = errorPrefix(result)
  if (name === "exec" || name === "python_runner") {
    const run = parseRunResult(result)
    if (run) {
      const failed = Boolean(flagged || run.failed || (run.exitCode !== 0 && run.exitCode !== undefined))
      const body = [run.stdout, run.stderr].filter((s) => s.trim()).join("\n")
      const error = run.error || (failed ? `exit ${run.exitCode ?? "?"}` : undefined)
      return { summary, failed, error, body }
    }
  }
  if (name === "web_search") {
    const hits = parseSearchHits(result)
    if (hits) {
      return { summary, failed: Boolean(flagged || prefix), error: prefix, body: "", hits }
    }
  }
  const failed = Boolean(flagged || prefix)
  if (name === "memory" || name === "skill_manage" || name === "skill_view") {
    const mem = parseMemoryResult(result)
    if (mem) {
      return {
        summary,
        failed: Boolean(flagged || !mem.success),
        error: mem.error || prefix,
        body: mem.body,
      }
    }
  }
  return { summary, failed, error: prefix, body: result }
}

function summariseMemoryCall(
  name: string,
  obj: Record<string, unknown> | undefined,
  raw: string,
): string {
  if (!obj) return flatten(raw)
  if (name === "skill_view") return asText(obj.name)
  const action = asText(obj.action)
  if (name === "skill_manage") {
    return [action, asText(obj.name)].filter(Boolean).join(" · ")
  }
  const preview = asText(obj.content) || asText(obj.old_text)
  return [action, preview].filter(Boolean).join(" · ")
}

function parseMemoryResult(raw: string): { success: boolean; error?: string; body: string } | undefined {
  const obj = parseObject(raw)
  if (!obj || typeof obj.success !== "boolean") return undefined
  if (obj.success === false) {
    return { success: false, error: asText(obj.error) || "refused", body: memoryFailureBody(obj) }
  }
  const parts: string[] = []
  const usage = asText(obj.usage)
  if (usage) parts.push(usage)
  if (Array.isArray(obj.entries)) {
    for (const e of obj.entries) {
      const t = asText(e)
      if (t) parts.push(t)
    }
  }
  const name = asText(obj.name)
  if (name) parts.push(name)
  const body = asText(obj.body)
  if (body) parts.push(body)
  return { success: true, body: parts.join("\n") }
}

function memoryFailureBody(obj: Record<string, unknown>): string {
  const entries = Array.isArray(obj.current_entries)
    ? obj.current_entries.map(asText).filter(Boolean)
    : []
  if (entries.length) return entries.join("\n")
  const available = Array.isArray(obj.available)
    ? obj.available.map(asText).filter(Boolean)
    : []
  // Repeating error as the body is how a refused skill_view showed the same
  // line twice. Extra context (what is stored / which skills exist) is the
  // only thing the body is for.
  return available.join("\n")
}

function parseRunResult(raw: string): {
  stdout: string
  stderr: string
  failed: boolean
  exitCode?: number
  error?: string
} | undefined {
  const obj = parseObject(raw)
  if (!obj) return undefined
  if (!("stdout" in obj) && !("stderr" in obj) && !("exit_code" in obj) && !("failed" in obj)) {
    return undefined
  }
  const exitCode = asNumber(obj.exit_code)
  return {
    stdout: asText(obj.stdout),
    stderr: asText(obj.stderr),
    failed: obj.failed === true,
    exitCode,
    error: asText(obj.error) || undefined,
  }
}

function parseSearchHits(raw: string): SearchHit[] | undefined {
  const val = parseJSON(raw)
  if (val === undefined) return undefined
  const rows = Array.isArray(val)
    ? val
    : isRecord(val)
      ? arrayOf(val.results) ?? arrayOf(val.items) ?? arrayOf(val.data)
      : undefined
  if (!rows) return undefined
  const hits: SearchHit[] = []
  for (const row of rows) {
    if (!isRecord(row)) continue
    const title = asText(row.title) || asText(row.name)
    const url = asText(row.url) || asText(row.link) || asText(row.href)
    const snippet =
      asText(row.snippet) ||
      asText(row.summary) ||
      asText(row.body) ||
      asText(row.content) ||
      asText(row.description)
    if (!title && !url && !snippet) continue
    hits.push({ title: title || url, url, snippet })
  }
  return hits.length ? hits : undefined
}

function errorPrefix(raw: string): string | undefined {
  const t = raw.trim()
  if (/^error:/i.test(t)) return t.replace(/^error:\s*/i, "")
  return undefined
}

function parseObject(raw: string): Record<string, unknown> | undefined {
  const v = parseJSON(raw)
  return isRecord(v) ? v : undefined
}

function parseJSON(raw: string): unknown {
  try {
    return JSON.parse(raw)
  } catch {
    return undefined
  }
}

function arrayOf(v: unknown): unknown[] | undefined {
  return Array.isArray(v) ? v : undefined
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return Boolean(v) && typeof v === "object" && !Array.isArray(v)
}

function asText(v: unknown): string {
  if (typeof v === "string") return v.trim()
  if (typeof v === "number" || typeof v === "boolean") return String(v)
  return ""
}

function asNumber(v: unknown): number | undefined {
  if (typeof v === "number" && Number.isFinite(v)) return v
  if (typeof v === "string" && v.trim() && Number.isFinite(Number(v))) return Number(v)
  return undefined
}

function flatten(s: string): string {
  return s.replace(/\s+/g, " ").trim()
}

/** Tools whose successful result means the project's files on disk changed. */
export const MEMORY_WRITE_TOOLS = new Set(["memory", "skill_manage"])

/** A memory/skill tool result that actually wrote something. A duplicate add
 *  reports success with changed: false — that is not a write. */
export function memoryWriteLanded(result?: string): boolean {
  const obj = parseObject(result ?? "")
  if (!obj || obj.success !== true) return false
  return obj.changed !== false
}
